constraint/pkg/apis/templates/transform_test.go

```diff
@@ -0,0 +1,355 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package templates
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+)
+
+func TestAddPreserveUnknownFields(t *testing.T) {
+	trueBool := true
+	testCases := []struct {
+		name string
+		v    *apiextensionsv1beta1.JSONSchemaProps
+		exp  *apiextensionsv1beta1.JSONSchemaProps
+	}{
+		{
+			name: "nil properties",
+			v: &apiextensionsv1beta1.JSONSchemaProps{
+				Properties: nil,
+			},
+			exp: &apiextensionsv1beta1.JSONSchemaProps{
+				XPreserveUnknownFields: &trueBool,
+				Properties:             nil,
+			},
+		},
+		{
+			name: "no information",
+			v:    &apiextensionsv1beta1.JSONSchemaProps{},
+			exp: &apiextensionsv1beta1.JSONSchemaProps{
+				XPreserveUnknownFields: &trueBool,
+			},
+		},
+		{
+			name: "type object",
+			v: &apiextensionsv1beta1.JSONSchemaProps{
+				Type: "object",
+			},
+			exp: &apiextensionsv1beta1.JSONSchemaProps{
+				XPreserveUnknownFields: &trueBool,
+				Type:                   "object",
+			},
+		},
+		{
+			name: "type array",
+			v: &apiextensionsv1beta1.JSONSchemaProps{
+				Type: "array",
+			},
+			exp: &apiextensionsv1beta1.JSONSchemaProps{
+				XPreserveUnknownFields: &trueBool,
```

Comment:
This should have a schema defined for `items`, per the spec:

https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#specifying-a-structural-schema

Reply:
While this makes total sense, I'm not sure what in the spec actually spells out that `type: array` will fail if not accompanied by `items`.  There is this quote: 

>specifies a non-empty type (via type in OpenAPI) for the root, for each specified field of an object node (via properties or additionalProperties in OpenAPI) and for each item in an array node (via items in OpenAPI), with the exception of:
> - a node with x-kubernetes-int-or-string: true
> - a node with x-kubernetes-preserve-unknown-fields: true

But this seems to say that an _item_ in an array node needs to have a type or `x-kub...`, but not that `items` needs to exist.

I did some searching through the [OpenAPI Spec](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.0.md) for `items` and `array` to look for a clue but did not find one.  Is there something that you've seen?

The best I can find is on json-schema.org, in the [Understanding JSON Schema page](https://json-schema.org/understanding-json-schema/reference/array.html#items):

> By default, the elements of the array may be anything at all. However, it’s often useful to validate the items of the array against some schema as well. This is done using the items, additionalItems, and contains keywords.

Overall, it feels a little ambiguous to me. But, I'm happy to follow if you feel you understand what's implied.

Reply:
From line item 1 for specifying a structural schema:

`and for each item in an array node (via items in OpenAPI),`

Reply:
Some comments from chatting with max:

- if `type: array` and no `items`, add `items` with `x-kub...` inside of it
- if `type: array`, don't put `x-kub...` at that level
- if no `type: array`, add `x-kub...`, even if there is `items`